### PR TITLE
LUCENE-4056: Japanese Tokenizer (Kuromoji) cannot build UniDic dictionary

### PR DESCRIPTION
**Pull request by johtani**
# Description

UniDic has a bit different column and unk.def from ipadic.
Currently if we use UniDic, then we get some error for building dictionary.
We can build UniDic after changing build.xml. I left some comments in build.xml.

If we use UniDic, we should change stoptags.txt and also add an example for part of speech filter.
UniDic has different part of speech tags from ipadic.

# Solution

Added some logic for adjusting UniDic csv and unk.def.
And also changing build.xml for downloading and building UniDic.

# Tests

Unfortunately there is no tests for DictionaryBuilder.

# Checklist

Please review the following and check all that apply:

- [x] I have reviewed the guidelines for [How to Contribute](https://wiki.apache.org/solr/HowToContribute) and my code conforms to the standards described there to the best of my ability.
- [x] I have created a Jira issue and added the issue ID to my pull request title.
- [x] I am authorized to contribute this code to the ASF and have removed any code I do not have a license to distribute.
- [x] I have given Solr maintainers [access](https://help.github.com/en/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork) to contribute to my PR branch. (optional but recommended)
- [x] I have developed this patch against the `master` branch.
- [ ] I have run `ant precommit` and the appropriate test suite.
- [ ] I have added tests for my changes.
- [ ] I have added documentation for the [Ref Guide](https://github.com/apache/lucene-solr/tree/master/solr/solr-ref-guide) (for Solr changes only).

### CODE REVIEW

File: lucene/analysis/kuromoji/ivy.xml [hunk 1]
```diff
@@ -19,19 +19,23 @@
...
   <dependencies>
-    <dependency org="mecab" name="mecab-ipadic" rev="${/mecab/mecab-ipadic}" conf="ipadic"> 
+    <dependency org="mecab" name="mecab-ipadic" rev="${/mecab/mecab-ipadic}" conf="ipadic">
       <artifact name="ipadic" type=".tar.gz" url="https://jaist.dl.sourceforge.net/project/mecab/mecab-ipadic/2.7.0-20070801/mecab-ipadic-2.7.0-20070801.tar.gz"/>
     </dependency>
     <dependency org="mecab" name="mecab-naist-jdic" rev="${/mecab/mecab-naist-jdic}" conf="naist">
       <artifact name="mecab-naist-jdic" type=".tar.gz" url=" https://rwthaachen.dl.osdn.jp/naist-jdic/53500/mecab-naist-jdic-0.6.3b-20111013.tar.gz"/>
     </dependency>
+    <dependency org="mecab" name="mecab-unidic" rev="${/mecab/mecab-unidic}" conf="unidic">
+      <artifact name="unidic" type=".zip" url=" http://ja.osdn.net/frs/redir.php?m=iij&amp;f=unidic%2F58338%2Funidic-mecab-2.1.2_src.zip"/>
```
Comment (msokolov) [inline lucene/analysis/kuromoji/ivy.xml]: If we download external files we must somewhere include a license demonstrating that those files are positively licensed for re-use, ideally with some recognized open source license. Do you know what the status is of this source for unidic? What is osdn.net?

**Comment by msokolov** [comment]
I understand there are no tests. Could you at least state for the record that you were able to build both UNIDIC and IPADIC-based dictionaries after your change using `ant build-dict`?

File: lucene/analysis/kuromoji/ivy.xml (hunk 1 above)
Comment (mocobeta) [reply lucene/analysis/kuromoji/ivy.xml]: About licensing, UniDic is distributed under GPL/LGPL/BSD(3-clause) triple licenses: https://unidic.ninjal.ac.jp/back_number#unidic_bccwj 
According to my knowledge, the last one is compatible with ASF's policy.

**Comment by mocobeta** [comment]
> Unfortunately there is no tests for DictionaryBuilder.

Yes it's a problem for future maintenance, I think we may need some kind of validator for binary encoded dictionaries rather than directly writing unit tests for the builder.

File: lucene/analysis/kuromoji/ivy.xml (hunk 1 above)
Comment (msokolov) [reply lucene/analysis/kuromoji/ivy.xml]: OK, great. We should add a reference to the unidic license to lucene/NOTICE.txt then, where we have some statement about ipadic already, and possibly we should copy the license into lucene/licenses/ - I'm not sure if that's needed.

Comment (johtani) [reply lucene/analysis/kuromoji/ivy.xml]: Thanks @msokolov and  @mocobeta .
I'll add the dictionary info in NOTICE.txt.

**Comment by johtani** [comment]
At least, I should paste the messages build successful for Unidic and Ipadic.
And I will think tests for binary dictionaries.

**Comment by azagniotov** [comment]
Hello Team,

May I inquire where are we on this?

## TL;DR

In the meanwhile, I attempted and succeeded to build the following dictionaries. I am using the tweaks that @johtani added in his PR three years ago, plus a few minor tweaks of my own:
- [unidic-cwj-202302_full](https://clrd.ninjal.ac.jp/unidic_archive/2302/) (NINJAL)
- [unidic-cwj-3.1.1-full](https://clrd.ninjal.ac.jp/unidic_archive/cwj/3.1.1/) (NINJAL) (See **Caveat** below 👇🏼 )
- [unidic-mecab-2.1.2_src](https://clrd.ninjal.ac.jp/unidic_archive/cwj/2.1.2/) (NINJAL)
- [mecab-ipadic-2.7.0-20070801](https://taku910.github.io/mecab/)

See the attached screenshots displaying the generated files👇🏼 I have also tested the built dictionaries, see **Testing** section below 👇🏼  

Shall I try make a new PR under https://github.com/apache/lucene in order to get a conversation re-started on this? cc: @mocobeta 🙇🏼‍♀️ 

## Building 

### Caveat
RE: Building the [unidic-cwj-3.1.1-full](https://clrd.ninjal.ac.jp/unidic_archive/cwj/3.1.1/): 

1. I had to increase the the [length check of the baseForm](https://github.com/apache/lucene/blob/2a3e5ca07f5df1f5080b5cb54ff104b7924e99f3/lucene/analysis/kuromoji/src/java/org/apache/lucene/analysis/ja/dict/TokenInfoDictionaryEntryWriter.java#L138-L140) from `16` to `35`
2. I had to stop [throwing an exception for multiple entries of LeftID](https://github.com/apache/lucene/blob/2a3e5ca07f5df1f5080b5cb54ff104b7924e99f3/lucene/analysis/kuromoji/src/java/org/apache/lucene/analysis/ja/dict/TokenInfoDictionaryEntryWriter.java#L129-L131). Instead, I printed them for curiosity:

```
...
...
Multiple entries found for leftID=4644: existing=[動詞-一般,五段-ガ行,連用形-融合], new fullPOSData=[動詞-一般,五段-ガ行,仮定形-融合]
Multiple entries found for leftID=4644: existing=[動詞-一般,五段-ガ行,仮定形-融合], new fullPOSData=[動詞-一般,五段-ガ行,連用形-融合]
Multiple entries found for leftID=4644: existing=[動詞-一般,五段-ガ行,連用形-融合], new fullPOSData=[動詞-一般,五段-ガ行,仮定形-融合]
Multiple entries found for leftID=4644: existing=[動詞-一般,五段-ガ行,仮定形-融合], new fullPOSData=[動詞-一般,五段-ガ行,連用形-融合]
Multiple entries found for leftID=14172: existing=[動詞-一般,五段-バ行,仮定形-融合], new fullPOSData=[動詞-一般,五段-バ行,連用形-融合]
Multiple entries found for leftID=1121: existing=[動詞-一般,五段-ガ行,連用形-融合], new fullPOSData=[動詞-一般,五段-ガ行,仮定形-融合]
Multiple entries found for leftID=1121: existing=[動詞-一般,五段-ガ行,仮定形-融合], new fullPOSData=[動詞-一般,五段-ガ行,連用形-融合]
Multiple entries found for leftID=4644: existing=[動詞-一般,五段-ガ行,連用形-融合], new fullPOSData=[動詞-一般,五段-ガ行,仮定形-融合]
Multiple entries found for leftID=4644: existing=[動詞-一般,五段-ガ行,仮定形-融合], new fullPOSData=[動詞-一般,五段-ガ行,連用形-融合]
Multiple entries found for leftID=5187: existing=[動詞-一般,五段-ガ行,連用形-融合], new fullPOSData=[動詞-一般,五段-ガ行,仮定形-融合]
Multiple entries found for leftID=5187: existing=[動詞-一般,五段-ガ行,仮定形-融合], new fullPOSData=[動詞-一般,五段-ガ行,連用形-融合]
Multiple entries found for leftID=8269: existing=[動詞-一般,五段-ガ行,連用形-融合], new fullPOSData=[動詞-一般,五段-ガ行,仮定形-融合]
Multiple entries found for leftID=8269: existing=[動詞-一般,五段-ガ行,仮定形-融合], new fullPOSData=[動詞-一般,五段-ガ行,連用形-融合]
Multiple entries found for leftID=9019: existing=[動詞-一般,五段-バ行,連用形-融合], new fullPOSData=[動詞-一般,五段-バ行,仮定形-融合]
Multiple entries found for leftID=9019: existing=[動詞-一般,五段-バ行,仮定形-融合], new fullPOSData=[動詞-一般,五段-バ行,連用形-融合]
Multiple entries found for leftID=11995: existing=[動詞-一般,五段-ガ行,連用形-融合], new fullPOSData=[動詞-一般,五段-ガ行,仮定形-融合]
Multiple entries found for leftID=11995: existing=[動詞-一般,五段-ガ行,仮定形-融合], new fullPOSData=[動詞-一般,五段-ガ行,連用形-融合]
Multiple entries found for leftID=7343: existing=[動詞-一般,五段-ガ行,連用形-融合], new fullPOSData=[動詞-一般,五段-ガ行,仮定形-融合]
Multiple entries found for leftID=7343: existing=[動詞-一般,五段-ガ行,仮定形-融合], new fullPOSData=[動詞-一般,五段-ガ行,連用形-融合]
Multiple entries found for leftID=9019: existing=[動詞-一般,五段-バ行,連用形-融合], new fullPOSData=[動詞-一般,五段-バ行,仮定形-融合]
Multiple entries found for leftID=9019: existing=[動詞-一般,五段-バ行,仮定形-融合], new fullPOSData=[動詞-一般,五段-バ行,連用形-融合]
Multiple entries found for leftID=13457: existing=[動詞-一般,上一段-ガ行,連体形-撥音便], new fullPOSData=[動詞-一般,上一段-ガ行,仮定形-融合]
...
...
```

### Gradle commands

The following has been performed on the fresh clone of https://github.com/apache/lucene:

My build command leveraged the new Gradle setup and the [DictionaryBuilder](https://github.com/apache/lucene/blob/main/lucene/analysis/kuromoji/src/java/org/apache/lucene/analysis/ja/dict/DictionaryBuilder.java) JavaDoc comment about how to do it.

I added in `lucene/analysis/kuromoji/build.gradle` a default `run` task from the Gradle's application plugin `apply plugin: 'application'`:
```
application {
  mainModule = 'org.apache.lucene.analysis.kuromoji' // name defined in module-info.java
  mainClass = 'org.apache.lucene.analysis.ja.dict.DictionaryBuilder'
}
```

My shell Gradle commands to build the dictionaries are as follows. The commands were executed under the root directory `lucene`, where the `gradlew` file is.

#### Building unidic-cwj-202302_full command
```
./gradlew -p lucene/analysis/kuromoji run --args='unidic "/Users/azagniotov/Downloads/unidic-cwj-202302_full" "/Users/azagniotov/Downloads/unidic-cwj-202302_full/lucene-kuromoji-built" "UTF-8" false'
```

#### Building unidic-cwj-3.1.1-full command
```
./gradlew -p lucene/analysis/kuromoji run --args='unidic "/Users/azagniotov/Downloads/unidic-cwj-3.1.1-full" "/Users/azagniotov/Downloads/unidic-cwj-3.1.1-full/lucene-kuromoji-built" "UTF-8" false'
```

#### Building unidic-mecab-2.1.2_src command
```
./gradlew -p lucene/analysis/kuromoji run --args='unidic "/Users/azagniotov/Downloads/unidic-mecab-2.1.2_src" "/Users/azagniotov/Downloads/unidic-mecab-2.1.2_src/lucene-kuromoji-built" "UTF-8" false'
```

#### Building mecab-ipadic-2.7.0-20070801 command
```
./gradlew -p lucene/analysis/kuromoji run --args='ipadic "/Users/azagniotov/Downloads/mecab-ipadic-2.7.0-20070801" "/Users/azagniotov/Downloads/mecab-ipadic-2.7.0-20070801/lucene-kuromoji-built" "EUC-JP" false'
```

## Testing

The built dictionaries were tested using the following Japanese strings: 

- `"にじさんじ"`
- `"ちいかわ"`
- `"桃太郎電鉄"`
- `"聖川真斗"`

The dictionaries metadata (dictionary by dictionary) were places under the [lucene/analysis/kuromoji/src/resources/org/apache/lucene/analysis/ja/dict](https://github.com/apache/lucene/tree/main/lucene/analysis/kuromoji/src/resources/org/apache/lucene/analysis/ja/dict) and a few unit test cases were added: 

#### Existing default dictionary
```
assertAnalyzesTo(analyzerNoPunct, "にじさんじ", new String[] {"に", "じさ", "ん", "じ"}, new int[] {1, 1, 1, 1});
assertAnalyzesTo(analyzerNoPunct, "ちいかわ", new String[] {"ちい", "か", "わ"}, new int[] {1, 1, 1});
assertAnalyzesTo(analyzerNoPunct, "桃太郎電鉄", new String[] {"桃太郎", "電鉄"}, new int[] {1, 1});
assertAnalyzesTo(analyzerNoPunct, "聖川真斗", new String[] {"聖川", "真", "斗"}, new int[] {1, 1, 1});
```

#### Built unidic-cwj-202302_full 

I need to increase memory before running the tests, as `ConnectionCosts.dat` is ~700MB

```
assertAnalyzesTo(analyzerNoPunct, "にじさんじ", new String[] {"にじ", "さん", "じ"}, new int[] {1, 1, 1});
assertAnalyzesTo(analyzerNoPunct, "ちいかわ", new String[] {"ちい", "かわ"}, new int[] {1, 1});
assertAnalyzesTo(analyzerNoPunct, "桃太郎電鉄", new String[] {"桃", "桃太郎", "太郎", "電鉄"}, new int[] {1, 0, 1, 1});
assertAnalyzesTo(analyzerNoPunct, "聖川真斗", new String[] {"聖", "川", "真", "斗"}, new int[] {1, 1, 1, 1});
```

#### Built unidic-cwj-3.1.1-full
```
assertAnalyzesTo(analyzerNoPunct, "にじさんじ", new String[] {"にじ", "さ", "ん", "じ"}, new int[] {1, 1, 1, 1});
assertAnalyzesTo(analyzerNoPunct, "ちいかわ", new String[] {"ちい", "かわ"}, new int[] {1, 1});
assertAnalyzesTo(analyzerNoPunct, "桃太郎電鉄", new String[] {"桃太郎", "電鉄"}, new int[] {1, 1});
assertAnalyzesTo(analyzerNoPunct, "聖川真斗", new String[] {"聖", "川", "真斗"}, new int[] {1, 1, 1});
```

#### Built unidic-mecab-2.1.2_src
```
assertAnalyzesTo(analyzerNoPunct, "にじさんじ", new String[] {"にじ", "さん", "じ"}, new int[] {1, 1, 1});
assertAnalyzesTo(analyzerNoPunct, "ちいかわ", new String[] {"ちい", "か", "わ"}, new int[] {1, 1, 1});
assertAnalyzesTo(analyzerNoPunct, "桃太郎電鉄", new String[] {"桃", "桃太郎", "太郎", "電鉄"}, new int[] {1, 0, 1, 1});
assertAnalyzesTo(analyzerNoPunct, "聖川真斗", new String[] {"聖", "川", "真", "斗"}, new int[] {1, 1, 1, 1});
```

#### Built mecab-ipadic-2.7.0-20070801
```
 assertAnalyzesTo(analyzerNoPunct, "にじさんじ", new String[] {"に", "じさ", "ん", "じ"}, new int[] {1, 1, 1, 1});
 assertAnalyzesTo(analyzerNoPunct, "ちいかわ", new String[] {"ちい", "か", "わ"}, new int[] {1, 1, 1});
 assertAnalyzesTo(analyzerNoPunct, "桃太郎電鉄", new String[] {"桃太郎", "電鉄"}, new int[] {1, 1});
 assertAnalyzesTo(analyzerNoPunct, "聖川真斗", new String[] {"聖川", "真", "斗"}, new int[] {1, 1, 1});
```


## Screenshots

#### Built unidic-cwj-202302_full
Screen Shot 2023-08-21 at 16 26 06

#### Built unidic-mecab-2.1.2_src
Screen Shot 2023-08-21 at 21 25 11


#### Built mecab-ipadic-2.7.0-20070801
Screen Shot 2023-08-21 at 17 36 04

**Comment by msokolov** [comment]
Yes if you want to revive the discussion, please move the PR over to the lucene git repo. I'm a little unclear on the future of this though. So far it's a pretty expert feature. To use it you have to edit the gradle build script and rebuild Lucene, right?

**Comment by azagniotov** [comment]
> Yes if you want to revive the discussion, please move the PR over to the lucene git repo. I'm a little unclear on the future of this though. So far it's a pretty expert feature. To use it you have to edit the gradle build script and rebuild Lucene, right?

@msokolov hello! 

Thank you. Yes, I figured that PR will need to be exported eventually under the right repo, thus, I went ahead and did it: 
- https://github.com/apache/lucene/pull/12517

To use this feature, yes, Lucene analysis Kuromoji JAR will have to be rebuilt after rebuilding a new dictionary. For example, while under the root lucene/ directory where the `gradlew` file is:
 
1. Run the: [./gradlew compileUnidic](https://github.com/apache/lucene/pull/12517/commits/bc9d2e3bb38d096ad193ce41e46c644b4c5673f9) (added in my PR). This will download and compile [unidic 3.1.1](https://github.com/apache/lucene/pull/12517/commits/bc9d2e3bb38d096ad193ce41e46c644b4c5673f9#diff-6bc685fcad3afe3dbc3692590a442b98efa1ae2912edc332fecbf8c8792b9333R142-R145) and will put the generated `*.dat` files under [the resources directory ](https://github.com/apache/lucene/tree/main/lucene/analysis/kuromoji/src/resources/org/apache/lucene/analysis/ja/dict)of the kuromoji package
2. To create a new Lucene analysis Kuromoji JAR with the new `*.dat` files, just run `./gradlew assemble`.  

#### To extend this example a little further, more specifically - using this together with Solr
My aforementioned [PR under Lucene](https://github.com/apache/lucene/pull/12517) adds the new changes using the latest state of code from the `main` branch.

If you are running Solr in Docker, then the PR changes must be adapted to the Lucene v9.7.0 branch, which is what the latest Solr is running on. This is something that I have done when I was rebuilding and testing the Lucene library. I have a patch [lucene_9.7.0_kuromoji_unidic_3_compatibility](https://gist.githubusercontent.com/azagniotov/8ed7d2d627199cd4365ae979aabd9575/raw/99e690f6c2bc2d059a7b416ee236ba546637a8ce/lucene_9.7.0_kuromoji_unidic_3_compatibility.patch) that adapts the PR changes to the v9.7.0 branch.

Thus, doing the above steps inside the Dockerfile, you could then replace the default Solr's `lucene-analysis-kuromoji-9.7.0` JAR after rebuilding it:

```
...
...
ENV SOLR_JAVA_MEM="-Xms2g -Xmx2g"
ENV SOLR_LIB_HOME=/opt/solr/server/solr-webapp/webapp/WEB-INF/lib

USER root
RUN rm $SOLR_LIB_HOME/lucene-analysis-kuromoji-*.jar
COPY ./lucene/analysis/kuromoji/build/libs/lucene-analysis-kuromoji-9.7.0-SNAPSHOT.jar $SOLR_LIB_HOME/
...
...
USER solr
```
